cabal-install/Distribution/Client/ProjectPlanning.hs

```diff
@@ -1731,25 +1731,31 @@ defaultSetupDeps compiler platform pkg =
         Just $
         [ Dependency depPkgname anyVersion
         | depPkgname <- legacyCustomSetupPkgs compiler platform ] ++
-        -- The Cabal dep is slightly special:
-        --  * we omit the dep for the Cabal lib itself (since it bootstraps),
-        --  * we constrain it to be less than 1.23 since all packages
-        --    relying on later Cabal spec versions are supposed to use
-        --    explit setup deps. Having this constraint also allows later
-        --    Cabal lib versions to make breaking API changes without breaking
-        --    all old Setup.hs scripts.
         [ Dependency cabalPkgname cabalConstraint
         | packageName pkg /= cabalPkgname ]
         where
-          cabalConstraint   = orLaterVersion (PD.specVersion pkg)
+          -- The Cabal dep is slightly special:
+          -- * We omit the dep for the Cabal lib itself, since it bootstraps.
+          -- * We constrain it to be >= 1.18 < 2
+          --
+          cabalConstraint   = orLaterVersion cabalCompatMinVer
+                                `intersectVersionRanges`
+                              orLaterVersion (PD.specVersion pkg)
                                 `intersectVersionRanges`
                               earlierVersion cabalCompatMaxVer
-        -- TODO/FIXME: turns out that constraining to less than 1.23 causes
-        --             problems with GHC8 as there's too many important packages
-        --             with Custom build-type, for which there wouldn't be any
-        --             install-plan (as GHC8 requires Cabal-1.24+). So let's
-        --             set an implicit upper bound `Cabal < 2` instead.
+          -- The idea here is that at some point we will make significant
```

Comment:
I guess we can just set the max bound to <= 1.22, since 1.24 already includes changes that break some setup scripts (see https://github.com/haskell/cabal/issues/3199#issuecomment-206780736).